### DISCUSSION TITLE
How does the GUI detect board type?

### DISCUSSION THREAD

**Discussion by letGitFoxy** [TeamOpenFIRE/OpenFIRE-App#22] [Q&A]
Hi there. Im in the process of designing a custom board for a project and I would like to know how the software detects what board is being used? I am planning to clone the original Pico but using a slightly different model for the flash memory and few other minor components. Thanks for your time.

**Comment by SeongGino** [comment]
Hiya! It's a bit of a process, but I'll lay it out to be as straightforward as possible.

On the firmware side, boards are defined with precompiler definitions listed at the top of https://github.com/TeamOpenFIRE/OpenFIRE-Firmware/blob/OpenFIRE-dev/libraries/OpenFIREPosition/OpenFIREBoard.h as strings corresponding to the board name - for custom boards, you'll want to add it to the top above the first main board (`ADAFRUIT_ITSYBITSY...`) so that it has priority, and make an `OPENFIRE_BOARD` definition for the shorthand name of your custom board. Any default pin layouts are defined as overrides under `SamcoEnhanced/SamcoPreferences.cpp`->`LoadPresets()` and `PresetCam()` (the main sketch code handles I2C pair negotiation). That's it for the fw part, though you'll also want to update the GitHub workflow file to add your board to the list for automatic builds (using whatever the base board profile, likely `rp2040:rp2040:generic` for the arduino-cli call).

On the App side, you'll want to modify these bits:
 - `constants.h`:
   - Add the board shorthand to `enum boardTypes_e` (preferably below `vccgndYD`, above `presetBoardsCount`)
   - If your board has two or more layouts, add it to `const uint8_t boardCustomPresetsCount[presetBoardsCount]`, making sure it follows the order of `enum boardTypes_e` above, and then use the existing *QStringList* and *int8_t 2D arrays* as template for your own alternative layouts - otherwise, just make the layout count for your board *0*.
   - Create your default layout, using the other `const boardLayout_t boardNameLayout[]`s as a template, setting default mappings for all 30 GPIOs.
 - Drop your board's SVG (graphical representation, must be vertical) to `boardPics` and add your board SVG to `vectors.qrc`, following the pre existing ones as a guideline.
 - `guiWindow.cpp`: For the most part, it's all "add your board to instances where `switch(board.type)` is used."
   - Add your board to the board string comparison branch (`if buffer[3] == "boardName" { board.type = boardName; } else ...`); again, preferably *below* `vccgndYD` and *above* the `generic` fallback.
   - Add your board to the default layouts switch case in `guiWindow::BoxesUpdate()`->`switch(board.type)`
   - Add your board's full name (the one the user sees in the UI) to the name assigner switch case in `PrettifyName()`->`switch(board.type)`
   - If your board has alternative layouts, it should be added to `guiWindow::BoxesFill()`->`switch(board.type)` and `guiWindow::on_presetsBox_currentIndexChanged(int index)`->`switch(board.type)`
   - Finally, the big mfer: adding your board layout to the layouts renderer switch case in `guiWindow::on_comPortSelector_currentIndexChanged(int index)`->`switch(board.type)`, using the preexisting ones as a template. Qt doesn't really have a very... *elegant* way of making elements appear/disappear without wanton destruction, so again, use the existing boards as a template.

Then just pull the changes to both repos and it will be reviewed accordingly.

I will be the first to admit that this isn't the prettiest way of adding content, as most of this was made when I was the one initially adding embedded programmable multi board support and the App code is, frankly, a bit of a hot mess - if anyone knows of a better way of doing this, contributions are of course accepted.
